### PR TITLE
beakerlib 1.30 (new formula)

### PR DESCRIPTION
**Pull request by LecrisUT**
This is a dependency for `tmt` which I will also add shortly. Basically these are test-frameworks which aim to make reproducible tests locally, in CI, in downstream packaging etc.

---

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/b/beakerlib.rb
Comment (SMillerDev) [inline Formula/b/beakerlib.rb:15]: Can this be
```suggestion
      source #{share}/beakerlib/beakerlib.sh || exit 1
```

**Comment by SMillerDev** [comment]
Wait until it's released

**Comment by SMillerDev** [comment]
You can still install things locally and host the formula in a personal tap, but in Homebrew core new formulae should not require patches.

**Comment by LecrisUT** [comment]
Sure, but how can we get a CI that would run on MacOS environments on the personal tap, or how to setup homebrew fork/tap to run the CI without the patch check there, etc.?

**Comment by SMillerDev** [comment]
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

**Comment by LecrisUT** [comment]
Ok, that did the trick. Not sure how to design the tap to be like a staging tap for upstream homebrew, e.g.:
- Having only smoke tests run on `homebrew-core`, but on the tap run a full test-suite
- Running the tap's CI via `workflow_run` using the upstream's git repo pointing to `main`, `PR` etc. 

Other automations like creating PRs to homebrew-core, sanitizing formulas before PR, I think I can handle.

**Comment by SMillerDev** [comment]
> Having only smoke tests run on homebrew-core, but on the tap run a full test-suite

Taps and core are completely separate so you can run whatever commands you like in it.
> Running the tap's CI via workflow_run using the upstream's git repo pointing to main, PR etc.

Don't think I know of any project doing this since for most Homebrew is just one of the many running cmake/meson/whatever. Sounds interesting though.
> Other automations like creating PRs to homebrew-core

There are GitHub actions that already exist for this, that should make this fairly easy.

File: Formula/b/beakerlib.rb
Comment (MikeMcQuaid) [inline Formula/b/beakerlib.rb:18]: This feels like it should be an option in the upstream `Makefile` rather than requiring us to patch it. Even better would be to avoid needing these GNU dependencies on macOS.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: Unfortunately that is tricky because this package is a `bash` based testing-framework so it uses a lot of linux specific commands, not only for the build, but for the runtime execution as well, i.e. the installed `.sh` files make use of extended `getopt` functions. This is also complicated by the fact that this must run consistently after reboot as well.

The replacement is overzealous, in that it replaces comments as well, but as for the functionality, it should work more reliably with such replacement.

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: As-is: this software does not seem to be ported to macOS.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: > As-is: this software does not seem to be ported to macOS.

Could you elaborate? Of course the functionality is not completely equivalent, but as long as the basic functionality like the `rlAssert` and journaling work, that is already a good enough first step for the `tmt` package integration. That I am already testing in the smoke-test included.

Having a reference formula would also help upstream expand the support of MacOS.

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: > Could you elaborate?

If the software requires `inreplace` like this to work on macOS: it is not ported to macOS.

You could argue further that if it requires GNU tools like this it's not ported but: that's not blocking this PR.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: Looking at the code implementation: https://github.com/search?q=repo%3Abeakerlib%2Fbeakerlib%20getopt&type=code

would you have a suggestion on how to make it macOS portable?

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: > I am not sure what you mean here, the upstream project does various detections to check for GNU vs BSD `getopt`

Why does the formula need to depend on/use `gnu-getopt` then?

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: Because the runtime shell scripts that are used by the user, e.g. `source #{share}/beakerlib/beakerlib.sh`, they **require** enhanced `getopt`. The checks inside the script check that `getopt` is enhanced, and if that fails, the whole script throws an error message. This happens both at the install stage, where some of these scripts are used internally, as well as at runtime when these are used by the user. So to get the **enhanced** `getopt` the `gnu-getopt` formula is needed with replacement of the `getopt` taken from `PATH` with the fixed absolute path to `gnu-getopt`

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: Ok, I see, thanks.

My numbered list from above still applies here.

Comment (sopos) [reply Formula/b/beakerlib.rb:18]: I can imagine it to be a part of the installation process in the Makefile (some `sed` processing).
I would rather keep the shell scripts untouched as there would be higher chance of direct usage in future development which would lead to a need of subsequent fixes afterwards.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: > @Bo98
> A PATH wrapper/configuration seems like the simplest approach here?

If you mean adding to `PATH` for the `.rb` formula, that wouldn't work because it needs to be made available after the installation for users that call beakerlib via `source .../beakerlib.sh`.

If you mean altering `PATH` within the scripts, we are just as well back to square one with the `sed`/`inreplace`

> @sopos 
> I can imagine it to be a part of the installation process in the Makefile (some sed processing).
> I would rather keep the shell scripts untouched as there would be higher chance of direct usage in future development which would lead to a need of subsequent fixes afterwards.

The idea for this is to add a configurable redirection to the actual `getopt` command, that will be `sed`-ed by homebrew, and otherwise left intact in order for the scripts to be executable in place, e.g. for the `logging.sh`
```patch
echo "${__INTERNAL_SOURCED}" | grep -qF -- " ${BASH_SOURCE} " && return || __INTERNAL_SOURCED+=" ${BASH_SOURCE} "

+ declare GETOPT_CMD=getopt

export __INTERNAL_DEFAULT_SUBMIT_LOG=__INTERNAL_FileSubmit

...

rlFileSubmit() {
-    GETOPT=$(getopt -o s: -- "$@" 2> >(while read -r line; do rlLogError "$FUNCNAME: $line"; done))
+    GETOPT=$(GETOPT_CMD -o s: -- "$@" 2> >(while read -r line; do rlLogError "$FUNCNAME: $line"; done))
    eval set -- "$GETOPT"
```
It would require yet another release to be accepted for homebrew, but we can experiment on the tap with patches to see if everything works fine

@SMillerDev @MikeMcQuaid We would appreciate accumulating some more feedback so that we don't have to cut new releases upstream this often.

As far as the `1.30` release is concerned, the `inreplace` does not do anything surprising (assuming it is correctly case-sensitive, which can be quickly sanity checked from the bottles), and it can be later refined on later release

Comment (Bo98) [reply Formula/b/beakerlib.rb:18]: > If you mean altering PATH within the scripts, we are just as well back to square one with the sed/inreplace

Is there multiple entry points for the scripts or does everything start with `beakerlib.sh`? If it's the latter then wrapping that with a PATH configuration is simple enough.

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: > I can imagine it to be a part of the installation process in the Makefile (some sed processing).

This makes sense to me too 👍🏻 

> It would require yet another release to be accepted for homebrew, but we can experiment on the tap with patches to see if everything works fine

This can be merged before the patches are in a release as long as they are merged upstream first.

Comment (sopos) [reply Formula/b/beakerlib.rb:18]: > This can be merged before the patches are in a release as long as they are merged upstream first.

The beakerlib upstream release quite cheap operation. If you provide reasonable PR, it can be merged and v1.30.1 can be done in a heart beat.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: Adding the `sed` replacement within the upstream `Makefile`:
- PR : https://github.com/beakerlib/beakerlib/pull/172
- Run in third-party tap: https://github.com/LecrisUT/homebrew-tmt/pull/1

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: @MikeMcQuaid The upstream patches were merged how do you want to proceed with this?

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb:18]: Thanks @LecrisUT! Apply the merged PR with a `patch do` block here and comment saying what it does and when it can be removed and adapt the `install` accordingly.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb:18]: :+1: How do I deal with the CI breaking down and not continuing to the test stage after it detects a `patch`? Can I link `patch do` to the [PR patch url](https://patch-diff.githubusercontent.com/raw/beakerlib/beakerlib/pull/172.patch)?

**Comment by LecrisUT** [comment]
Note, beware when updating this package, `spglib`, etc. https://github.com/scikit-build/scikit-build-core/issues/695. The `sha256sum` is changing after a commit is added: https://github.com/scikit-build/scikit-build-core/issues/695

File: Formula/b/beakerlib.rb
Comment (LecrisUT) [reply Formula/b/beakerlib.rb]: Still can't get around it ><:
```
==> FAILED
Full audit --formula beakerlib --online --new output
  beakerlib
    * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
  Error: 1 problem in 1 formula detected.
```

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb]: Yes. Check how other formulae use patches/PR URLs.

Comment (MikeMcQuaid) [reply Formula/b/beakerlib.rb]: @LecrisUT That's fine, we can ignore/override that.

Comment (LecrisUT) [reply Formula/b/beakerlib.rb]: Well that would need to be done in this PR somehow in order to run the CI for all macOS variants to confirm that everything is working

File: audit_exceptions/provided_by_macos_depends_on_allowlist.json
Comment (MikeMcQuaid) [inline audit_exceptions/provided_by_macos_depends_on_allowlist.json:4]: Please remove this. We can allow the dependency instead.

Comment (LecrisUT) [reply audit_exceptions/provided_by_macos_depends_on_allowlist.json:4]: This was discussed in https://github.com/Homebrew/homebrew-core/issues/165752
@Bo98 Can you comment on this? By the nature of the `gnu-getopt` package, all new packages will fail at `brew audit --online --new`, because the dependent package are **supposed** to use `uses_from_macos`. For any dependents of `gnu-getopt` this seems like an unnecessary barrier for new packages.

What I proposed in the issue is to cherry-pick this commit, but I haven't heard any feedback on if I should go ahead and make a PR about it since then.

Comment (MikeMcQuaid) [reply audit_exceptions/provided_by_macos_depends_on_allowlist.json:4]: > For any dependents of gnu-getopt this seems like an unnecessary barrier for new packages.

It's not, it just requires a maintainer to override the check before merging. This exists as a "human please check this".

**Comment by LecrisUT** [comment]
Well, I've rebased if someone wants to do the magic of making the CI work :shrug:

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks @LecrisUT!

**Pull request merged**
